icaruscode/CRT/CRTTPCTruthEffMCKalman_module.cc

```diff
@@ -0,0 +1,1721 @@
+		pfp_pdg = pfp->PdgCode();
+		pfp_isprimary = pfp->IsPrimary();
+
+	    //Find T0 for PFParticle
+	    auto t0 = fot0pandora.at(pfp.key());
+	    if (t0.isNonnull()){
+	      track_t0 = t0->Time();
+	      is_catcross = true;
+	      catcross_x = DBL_MAX;	catcross_y = DBL_MAX;	catcross_z = DBL_MAX;
+	      getCatCrossXYZ(*trackList[track_i], catcross_x, catcross_y, catcross_z);
+//	      t0 = track_t0;   //Get T0
+	    }//end if (!t0.empty())
+	  }//end if (!pfps.empty())
+
+	  crt_tpc_dca = DBL_MAX; crt_trueID = INT_MAX; track_trueID = INT_MAX; track_mother = INT_MAX; crttime = DBL_MAX;
+
+	  std::vector<art::Ptr<recob::Hit>> hits = findManyHits.at(trackList[track_i].key() /*->ID()*/);
+
+	  getTickRange(hits,tpc_mintick,tpc_maxtick);
+
+	  is_ancatcross = false;
+	  if(tpc_maxtick - tpc_mintick>=2300) is_ancatcross=true;
+
+
+	  if (hits.empty()) { std::cout << "hits object is EMPTY, CONTINUING\n";  continue;}
+	  //Get Truth info if applicable:
+//		track_trueID = RecoUtils::TrueParticleIDFromTotalRecoHits(clockData,hits,false);
+		track_trueID = RecoUtils::TrueIDOfTrack(clockData,hits,false);
+		GetAncestorID(track_trueID, track_mother, track_ancestor, track_motherlayers, particles);
+		trk_pdg = particles[track_trueID].PdgCode();
+		trk_ancestor_pdg = particles[track_ancestor].PdgCode();
+
+//		std::cout << "This track TrackID: " << track_trueID << "\nThis Track ancestor TrackID: " << track_ancestor << "\nThis Track PDG: " << trk_pdg << std::endl;
+
+	  //This can't easily be made into a function (with my coding skills, anyway) so I'm just gonna kludge a way of extracting any CRT 
+	  //Hits that are a "truth match" to the TPC ancestor ID found above
+	  num_crt_with_same_trueID = 0; num_crt_with_same_ancestor = 0; 
+	  crt_truthmatch_x.clear(); crt_truthmatch_y.clear(); crt_truthmatch_z.clear(); crt_truthmatch_time.clear();
+	  found_ancestor_match=false; found_trueID_match=false;
+
+	  for(auto &crtHit : crtHits){
+
+		int crt_trueID =  bt.TrueIdFromTotalEnergy(event, crtHit); int crt_mother = INT_MAX-5; int crt_ancestor=INT_MAX-5; int crt_motherlayer = INT_MAX-5;
+		GetAncestorID(crt_trueID, crt_mother, crt_ancestor, crt_motherlayer, particles);
+//		int this_crt_pdg = particles[crt_trueID].PdgCode();
+	
+		if(crt_trueID==track_trueID){ found_trueID_match=true; num_crt_with_same_trueID++; }
+		if(crt_ancestor==track_ancestor){ 
+			found_ancestor_match=true; 
+			num_crt_with_same_ancestor++; 
+			crt_truthmatch_x.push_back(crtHit.x_pos); 
+			crt_truthmatch_y.push_back(crtHit.y_pos); 
+			crt_truthmatch_z.push_back(crtHit.z_pos); 
+			double this_crtTime = double(crtHit.ts0() - 1600000)/1e3;
+			crt_truthmatch_time.push_back(this_crtTime);
+		}//end if(crt_ancestor==track_ancestor)
+
+
+	  }//End loop over crtHits
+
+
+//	  int const cryoNumber = hits[0]->WireID().Cryostat;
+
+	  matchCand_PCA closest = t0Alg.GetClosestCRTHit_PCA(detProp, *trackList[track_i], hits, crtHits,  m_gate_start_timestamp, fIsData);
+
+	  if(closest.dca >=0 ){
```

Comment:
It's not immediately clear to me if both of these if statements are necessary. When is the DCA negative?